src/vulkan-renderer/wrapper/device.cpp

```diff
@@ -0,0 +1,215 @@
+        device_extensions_wishlist.push_back(VK_EXT_DEBUG_MARKER_EXTENSION_NAME);
+    }
+#endif
+
+    std::vector<const char *> enabled_device_extensions;
+
+    for (auto device_extension_name : device_extensions_wishlist) {
+        if (availability_checks.has_device_extension(graphics_card, device_extension_name)) {
+            spdlog::debug("Device extension '{}' is available on this system.", device_extension_name);
+            enabled_device_extensions.push_back(device_extension_name);
+        } else {
+            throw std::runtime_error("Device extension " + std::string(device_extension_name) + " is not available on this system!");
+        }
+    }
+
+    VkPhysicalDeviceFeatures used_features = {};
+
+    // Enable anisotropic filtering.
+    used_features.samplerAnisotropy = VK_TRUE;
+
+    VkDeviceCreateInfo create_info = {};
+    create_info.sType = VK_STRUCTURE_TYPE_DEVICE_CREATE_INFO;
+    create_info.queueCreateInfoCount = static_cast<std::uint32_t>(queues_to_create.size());
+    create_info.pQueueCreateInfos = queues_to_create.data();
+    // Device layers were deprecated in Vulkan some time ago, essentially making all layers instance layers.
+    create_info.enabledLayerCount = 0;
+    create_info.ppEnabledLayerNames = nullptr;
+    create_info.enabledExtensionCount = static_cast<std::uint32_t>(enabled_device_extensions.size());
+    create_info.ppEnabledExtensionNames = enabled_device_extensions.data();
+    create_info.pEnabledFeatures = &used_features;
+
+    spdlog::debug("Creating physical device (graphics card interface).");
+
+    if (vkCreateDevice(graphics_card, &create_info, nullptr, &device) != VK_SUCCESS) {
+        throw std::runtime_error("Error: vkCreateDevice failed!");
+    }
+
+    // TODO: Assign internal Vulkan debug marker name to device and physical device.
+
+    spdlog::debug("Initialising GPU queues.");
+
+    spdlog::debug("Graphics queue family index: {}.", graphics_queue_family_index.value());
+    spdlog::debug("Presentation queue family index: {}.", present_queue_family_index.value());
+    spdlog::debug("Data transfer queue family index: {}.", transfer_queue_family_index.value());
+
+    // Setup the queues for presentation and graphics.
+    // Since we only have one queue per queue family, we acquire index 0.
+    vkGetDeviceQueue(device, present_queue_family_index.value(), 0, &present_queue);
+    vkGetDeviceQueue(device, graphics_queue_family_index.value(), 0, &graphics_queue);
+
+    // The use of data transfer queues can be forbidden by using -no_separate_data_queue.
+    if (use_distinct_data_transfer_queue) {
+        // Use a separate queue for data transfer to GPU.
+        vkGetDeviceQueue(device, transfer_queue_family_index.value(), 0, &transfer_queue);
+    }
+
+    spdlog::debug("Created device successfully.");
+}
+
+Device::~Device() {
+    vkDestroyDevice(this->device, nullptr);
+    this->device = VK_NULL_HANDLE;
+
+    queues_to_create.clear();
```

Comment:
This doesn't need to be cleared.

Reply:
Ah right. This came from copying old code.

Reply:
Done.